### ISSUE TITLE
conda build test phase does not test package which was built

### ISSUE THREAD

**Issue by msarahan** [conda/conda#2885]
@jjhelmus commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051)

The test phase of conda build will pick up packages with the same build string in the main conda `pkgs` directory which can result in a test which should fail to pass.  An easy method to observer this is to create a recipe which passed, build it, then modify it so that it should fail in the test phase.  If the build number has not changed the working package  will be installed during the test phase rather than the newly built package.

A tell-tale sign of this behavior is that the package is not listed as being downloaded during the test phase.

I am seeing this behavior on a OS X system with the follow conda info:

```
$ conda info
Current conda install:

             platform : osx-64
        conda version : 4.1.3
    conda-env version : 2.5.1
  conda-build version : 1.21.2
       python version : 3.5.1.final.0
     requests version : 2.10.0
     root environment : /Users/jhelmus/anaconda  (writable)
  default environment : /Users/jhelmus/anaconda
     envs directories : /Users/jhelmus/anaconda/envs
        package cache : /Users/jhelmus/anaconda/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/osx-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/osx-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /Users/jhelmus/.condarc
         offline mode : False
    is foreign system : False
```

---

@Korijn commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051#issuecomment-228721930)

`1.21.0` was the last version that did not have this bug. We've pinned the versions of conda, conda-build and conda-env across our whole system, many regressions are being introduced lately and it's... distracting, to say the least. :)

---

@msarahan commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051#issuecomment-228743126)

Thanks for the report.  This is a good excuse to finally have the test phase actually test the built module, rather than some strange series of "read metadata from built package, but then try to download package using that metadata."

@Korijn, my sincerest apologies for all the bugs I've introduced.  As part of #953, I have discovered that we only have about 45% test coverage (prior to that PR, most of our tests did not give us coverage information, because they run in subprocesses).  With that, breakage is somewhat inevitable. =(

I hope we'll have enough stability and new features to win you back soon.  Any tests you propose, whether as problem descriptions in issues like this one, as pseudocode, or as PR's, will be most helpful in assuring conda-build's stability going forward.

---

@Korijn commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051#issuecomment-228746997)

@msarahan Well, I'm married to conda already, so no need to win me over! I'm eagerly awaiting new versions to see if I can remove the version pins yet... :)

---

@jjhelmus commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051#issuecomment-228848290)

Minimal example which exhibits the broken behavior:

**meta.yaml**

``` yaml
package:
  name: foobar
  version: 1.0.0

build:
  number: 0
  script: echo "exit 0" > $PREFIX/bin/foobar; chmod +x $PREFIX/bin/foobar

test:
  commands:
    - foobar
```
- Run `conda build .` in a directory with this file, test phase should pass.
- Change "exit 0" to "exit 1", rerun `conda build .`, test phase should fail but will pass
- Change the build number to 1, rerun `conda build .`, test phase will now correctly fail.

Removing the foobar package from the package cache will also result in the correct behavior.

Some additional investigation leads me to believe that this is actually a bug in `conda` not `conda-build`.  Downgrading conda to version `4.0.9` removes this buggy behavior, upgrading to `4.1.0` and above causes the erroneous behavior.  Feel free to move this issue to the correct repository, ping me if you would like me to open an issue there.

---

@msarahan commented on [Mon Jun 27 2016](https://github.com/conda/conda-build/issues/1051#issuecomment-228849196)

Thanks @jjhelmus.  Considering that varying conda version changes behavior, I'm inclined to move this issue over there.  I will still be working to improve the way that packages are tested (more directly), but I think that's orthogonal to this issue.  @kalefranz @mcg1969, incoming!

**Comment by Korijn** [comment]
Nice detective work! I was having trouble reproducing.

**Comment by mcg1969** [comment]
The problem here is actually pretty simple to explain.

After the first run, `conda` "downloads" `foobar-1.0.0-0.tar.bz2` into its cache from the `conda-bld` channel. During the second run, _it's already in the cache_, so it doesn't think it needs to download it again.

The "problem" from conda-build's perspective is that conda is doing a better job of keeping track of the packages in its cache, and what channels it came from. This is absolutely the right thing for us to do for typical conda usage, but clearly it is problematic for conda-build. It doesn't know that the version of the tarball it has in its cache has expired, and I'm not sure it has any way of knowing this.

One simple solution would be to remove the tarball from `pkgs` before building the test environment.

**Comment by jjhelmus** [comment]
Personally I do not think any packages build using conda-build should be placed in the package cache but I can see the argument for this.  From what I've gathered conda now will only install packages from the cache so the problem may be a bit more tricky.

**Comment by mcg1969** [comment]
conda-build is using conda's standard channel capability to "reveal" its packages to conda proper. It would be interesting if we could find a way to specify that a particular channel is "ephemeral", and therefore to always assume that the cached packages are stale.

**Comment by ilanschnell** [comment]
I've had this problem too.  One solution would be to basically to a `conda install --force -n _test` before conda build is testing the package.  Another solution would be to remove the built package from the cache.

**Comment by Korijn** [comment]
My two cents: If this conda-build behavior is going to continue to depend on conda _implementation details_, there should at least be a very strong set of unit and regression tests.

Conda sees a lot of refactoring and thus I would strongly recommend depending on stable high-level features rather than low-level intricacies.

**Comment by ilanschnell** [comment]
I agree with @Korijn.  As the package cache is a conda implementation detail, it would be better to depend of the high-level feature `conda install --force -n _test`.

**Comment by mcg1969** [comment]
I'm not sure that `--force` is enough. That guarantees re-extraction, but not re-fetch. We will have to determine the proper external `conda` behavior is to ensure re-fetching, adding something if needed. Clearing the package cache might be the most expedient choice until then.

**Comment by ilanschnell** [comment]
`--force` does not re-fetch when the conda package is already in the cache and has the correct MD5 sum.  But since we just created the package, the hash will be fresh.

**Comment by mcg1969** [comment]
It doesn't check the hash unless you ask it to.

**Comment by mcg1969** [comment]
Actually, scratch that, I understand what you're saying here. Let's definitely make sure the MD5 is checked in this case, since we have that information.

**Comment by ilanschnell** [comment]
`--force` (by design) checks the MD5.  If not, it's a conda bug.

**Comment by mcg1969** [comment]
There have always been situations where the MD5 is skipped; but yes, let's definitely make sure that's not happening in this case.

**Comment by kalefranz** [comment]
I actually think this should be solved in 4.3.  Conda redownloads tarballs now if md5s don't match.  Fixing this issue actually ended up leading to this bug https://github.com/conda/conda/pull/5148, which I think we have fixed in 4.4 now with #5289.
